psh/wget/wget.c

```diff
@@ -0,0 +1,485 @@
+	if (fd < 0) {
+		printf("Failed\n");
+		freeaddrinfo(res);
+		return -1;
+	}
+
+	if (connect(fd, res->ai_addr, res->ai_addrlen) < 0) {
+		close(fd);
+		freeaddrinfo(res);
+		return -1;
+	}
+	printf("Connected\n");
+
+	freeaddrinfo(res);
+
+	return fd;
+}
+
+
+static int wget_request(void)
+{
+	char buf[256];
+	int bytes = 0, i = 0, ret;
+
+	bytes = snprintf(buf, sizeof(buf), "GET /%s HTTP/1.1\r\nHost: %s\r\nUser-Agent: Wget\r\n\r\n",
+		common.path, common.host);
+	if ((bytes < 0) || (bytes >= sizeof(buf))) {
+		fprintf(stderr, "wget: Request url too large!\n");
+		return -1;
+	}
+
+	while (i < bytes) {
+		ret = write(common.socket, buf + i, bytes - i);
+		if (ret <= 0) {
+			return -1;
+		}
+
+		i += ret;
+	}
+
+	if (common.debug) {
+		printf("--------Request--------\n%*s\n", bytes, buf);
+	}
+
+	return 0;
+}
+
+
+static size_t wget_file(size_t len)
+{
+	ssize_t ret;
+	size_t left = len;
+	size_t bytes;
+	int i = 0;
+
+	/* Save what is left in the buffer */
+	if (common.readptr < common.buf + common.len) {
+		bytes = common.buf + common.len - common.readptr;
+		if (fwrite(common.readptr, 1, bytes, common.outfile) != bytes) {
+			fprintf(stderr, "wget: Write failed\n");
+			return -1;
+		}
+		left -= bytes;
+	}
+
+	while (left > 0) {
+		ret = read(common.socket, common.buf, min(sizeof(common.buf), left));
+		if (ret <= 0) {
+			fprintf(stderr, "\nwget: read from socket failed!\n");
+			return -1;
+		}
+
+		left -= ret;
+
+		if (fwrite(common.buf, 1, ret, common.outfile) != ret) {
+			fprintf(stderr, "\nwget: Write failed\n");
+			return -1;
+		}
+		if (i == 0) {
+			printf("\rWritten: %8zu/%zu", len - left, len);
+			fflush(stdout);
+		}
+		i = (i + 1) % 8;
+	}
+	printf("\rWritten: %8zu/%zu\n", len - left, len);
+
+	return 0;
+}
+
+
+static int wget_buf(void)
+{
+	int ret;
+
+	/* Move not consumed part of the buffer */
+	common.len = common.len - (common.readptr - common.buf);
+	memmove(common.buf, common.readptr, common.len);
+	common.readptr = common.buf;
+
+	ret = read(common.socket, common.buf + common.len, sizeof(common.buf) - common.len);
+	if (ret <= 0) {
+		return -1;
+	}
+
+	common.len += ret;
+
+	return 0;
+}
+
+
+static char *wget_hdrnext(void)
+{
+	char *hdr = NULL, *tmp;
+
+	while (hdr == NULL) {
+		tmp = memchr(common.readptr, '\n', common.buf + common.len - common.readptr);
+		if (tmp == NULL) {
+			/* Carry on reading */
+			if (wget_buf() != 0) {
+				return NULL;
+			}
+			continue;
+		}
+
+		*(tmp - 1) = '\0';
+		hdr = common.readptr;
+		common.readptr = tmp + 1;
+	}
+
+	return hdr;
+}
+
+
+static size_t wget_parsehdrs(void)
+{
+	const char contenthdr[] = "content-length: ";
+	char *hdr = NULL, *endptr, *num;
+	size_t contentLen = 0;
+
+	/* Read headers until no more data is available or empty line is found */
+	do {
+		hdr = wget_hdrnext();
+		if (hdr == NULL) {
+			return -1;
+		}
+
+		if (strncasecmp(hdr, contenthdr, sizeof(contenthdr) - 1) == 0) {
+			num = strrchr(hdr, ' ');
```

Comment:
Ok, since I use `strrchr()` instead of `strchr()` I handle any number of spaces greater or equal to 1, but I'm not handling a case, where is not space at all. I'll fix that, thanks.